src/lib.rs

```diff
@@ -200,78 +237,247 @@ impl Default for ThreadCounters {
+
+        assert_eq!(allocator.count(), expected);
+
+        scope(|scope| {
+            for allocation in allocations {
+                scope.spawn(move |_scope| drop(allocation));
+            }
+        })
+        .unwrap();
+
+        assert_eq!(
+            allocator.count(),
+            AllocStats {
+                all_time: AllTimeAllocStats { dealloc: expected.all_time.alloc, ..expected.all_time },
+                since_last: IncrementalAllocStats { dealloc: expected.all_time.alloc, ..Default::default() },
+            }
+        );
+    }
 
     #[test]
-    fn display() {
-        let threads = (0..9)
-            .map(|idx| spawn(move || drop(Vec::<u8>::with_capacity(10000 * idx))))
-            .collect::<Vec<_>>();
-        threads.into_iter().for_each(|thread| thread.join().unwrap());
-
-        println!("{GLOBAL_ALLOCATOR:?}");
-        println!("{GLOBAL_ALLOCATOR}");
-        println!("{GLOBAL_ALLOCATOR:?}");
+    fn alloc_zeroed() {
+        let allocator = &Default::default();
+        let allocations = test_allocations(&allocator, AllocationHandle::new_zeroed, identity);
+        let expected = expected_counts(&allocations);
+
+        assert_eq!(allocator.count(), expected);
     }
 
     #[test]
-    fn count() {
-        let threads = (0..9)
-            .map(|idx| spawn(move || drop(Vec::<u8>::with_capacity(10000 * idx))))
-            .collect::<Vec<_>>();
-        threads.into_iter().for_each(|thread| thread.join().unwrap());
-
-        println!("{GLOBAL_ALLOCATOR:?}");
-        let AllocCounts { alloc, dealloc } = GLOBAL_ALLOCATOR.count();
-        println!("Total: alloc {alloc} dealloc {dealloc}");
-        println!("{GLOBAL_ALLOCATOR:?}");
+    fn realloc() {
+        let allocator = &Default::default();
+        let mut allocations = test_allocations(&allocator, AllocationHandle::new, identity);
+        let expected = expected_counts(&allocations);
+
+        assert_eq!(allocator.count(), expected);
+
+        scope(|scope| {
+            for allocation in &mut allocations {
+                scope.spawn(move |_scope| allocation.realloc(allocation.layout.size() * 2));
+            }
+        })
+        .unwrap();
+
+        let expected_2 = expected_counts(&allocations);
+
+        assert_eq!(
+            allocator.count(),
+            AllocStats {
+                all_time: AllTimeAllocStats {
+                    alloc: expected.since_last.alloc + expected_2.since_last.alloc,
+                    dealloc: expected.since_last.alloc,
+                    largest_alloc: expected_2.since_last.largest_alloc,
+                },
+                since_last: IncrementalAllocStats { dealloc: expected.since_last.alloc, ..expected_2.since_last }
+            }
+        );
+    }
+
+    unsafe impl GlobalAlloc for TestAlloc {
+        unsafe fn alloc(&self, layout: Layout) -> *mut u8 {
+            Box::into_raw(Box::new(Allocation { layout })) as *mut u8
+        }
```

Comment:
Oh wow, I had to figure out that this was **ignoring** the layout for the returned pointer. Can you put that in a comment to remind people not to write anything into these allocations?

(Also, may be worth commenting that this isn't actually a valid global allocator, since that would create a circular dependency.)

Reply:
Right. I'll add a comment. Since we're not actually using the memory, we don't actually need to allocate anything, but allocating _something_ using the system allocator let's us leverage it for double-free / leak detection. We save the layout in there purely to assert that the Layout upon dealloc matches, for verifying correctness.

Reply:
fixed in fecd6f2